src/azul/plugins/repository/canned/__init__.py

```diff
@@ -69,7 +73,105 @@
 log = logging.getLogger(__name__)
 
 
-class CannedSourceRef(SourceRef[SimpleSourceSpec, 'CannedSourceRef']):
+@attrs.frozen(kw_only=True)
+class CannedSourceSpec(SourceSpec):
+    url: str
+    ref: str
+    path: tuple[str, ...]
+    sep: ClassVar[str] = '::'
+
+    @classmethod
+    def parse(cls, spec: str) -> Self:
+        """
+        Construct an instance from its string representation, using the syntax
+        '{repo-url}::{ref}::{path}::{prefix}' where:
+        {repo-url} is the HTTPS URL of a GitHub repository's `.git` file. It
+                   must match the format https://github.com/USER/REPO.git
+        {ref} is a branch name, tag, or commit SHA
+        {path} is the path to the base of the staging area (e.g. 'tests/area1')
+        {prefix} is the common prefix and partition length (e.g. '/0')
+
+        >>> s = CannedSourceSpec.parse('https://github.com/user/repo.git::f0f0f0f0::dir/subdir::/0')
+        >>> s # doctest: +NORMALIZE_WHITESPACE
+        CannedSourceSpec(prefix=Prefix(common='', partition=0),
+                         url='https://github.com/user/repo.git',
+                         ref='f0f0f0f0',
+                         path=('dir', 'subdir'))
+
+        >>> s = CannedSourceSpec.parse('https://github.com/user/repo.git::f0f0f0f0::::42/0')
+        >>> s # doctest: +NORMALIZE_WHITESPACE
+        CannedSourceSpec(prefix=Prefix(common='42', partition=0),
+                         url='https://github.com/user/repo.git',
+                         ref='f0f0f0f0',
+                         path=())
+
+        >>> s = CannedSourceSpec.parse('https://github.com/foo.git::f0f0f0f0::::/0')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: ('Invalid GitHub repo URL', 'https://github.com/foo.git')
+        """
+        url, ref, path, prefix = cls._parse(spec)
+        self = cls(prefix=prefix,
+                   url=url,
+                   ref=ref,
+                   path=path)
+        assert spec == str(self), spec
+        return self
+
+    @classmethod
+    def _parse(cls, spec: str) -> tuple[str, str, tuple[str, ...], Prefix]:
+        require(spec.count(cls.sep) == 3, 'Invalid source specification', spec)
+        url, ref, path, prefix = spec.split(cls.sep)
+        require(url.startswith('https://'), 'Invalid GitHub repo URL', url)
+        require(url.endswith('.git'), 'Invalid GitHub repo URL', url)
+        require(len(furl(url).path.segments) == 2, 'Invalid GitHub repo URL', url)
```

Comment:
I'm always a bit nervous when we try to parse URLs using regular string functions. This is a bit more verbose, but I think it's also more semantically specific. What do you think?

```suggestion
        parsed_url = furl(url)
        require(parsed_url.scheme == 'https'), 'Invalid GitHub repo URL', url)
        require(parsed_url.pathstr.endswith('.git'), 'Invalid GitHub repo URL', url)
        require(len(parsed_url.path.segments) == 2, 'Invalid GitHub repo URL', url)
```

Reply:
`pathstr` is a deprecated property so instead I'll check for `.git` on the last path segment (`parsed_url.path.segments[-1]`).